Intrigue

jbw3 opened the pull request:
# Intrigue

I'm getting close to (but not quite) done with the Intrigue expansion, so I wanted to go ahead and push up a draft PR in case you had any feedback. The main thing I have left is to add functions to `OptimizedBotDecider` to handle the new cards.

Below are some changes worth noting:

## `Card` `actions`, `draw`, and `money`

The `Action` class has `actions`, `draw`, and `money` parameters to describe what the card does. I only filled these in when the card was guaranteed to give these values.
For example, I set `draw` to 2 for [Diplomat](http://wiki.dominionstrategy.com/index.php/Diplomat) because it always draws 2 cards, but I did not set `actions` since Diplomat only sometimes gives the player more actions. Let me know if you think I should do this differently.

## New `Decider` Functions

I added 5 new functions to the `Decider` protocol:

### `multiple_option_decision()`

This is used for many Intrigue cards and gives the player a choice between multiple options. It takes a list of strings as the options and returns an integer which is the index of the chosen option. I added a `Choice` enum to all classes that used this to make the code more readable when checking the integer value.

### `deck_position_decision()`

This is for [Secret Passage](http://wiki.dominionstrategy.com/index.php/Secret_Passage).

### `reveal_decision()`

This is for [Courtier](http://wiki.dominionstrategy.com/index.php/Courtier).

### `pass_decision()`

This is for [Masquerade](http://wiki.dominionstrategy.com/index.php/Masquerade).

### `name_card_decision()`

This is for [Wishing Well](http://wiki.dominionstrategy.com/index.php/Wishing_Well).

## Added `actions_played_this_turn`

I added this new variable to `Player` because [Conspirator](http://wiki.dominionstrategy.com/index.php/Conspirator) needs to know the number of actions a player has played on their turn. This is more complicated than counting the number of action on the player's playmat because if a player plays an action with Throne Room, that action is counted as being played twice even though there will only be one card on the playmat.

## Bridge and `get_cost()`

I had to make some changes to support [Bridge](http://wiki.dominionstrategy.com/index.php/Bridge) as it reduces the cost of all cards. I made `Card`'s `cost` member "private" (i.e. I changed it to `_cost`) and added a `get_cost()` member function to calculate a card's cost. `get_cost()` has `Player` and `Game` parameters so it can adjust the base cost based on any player or game state. I then added a `card_cost_reduction` member to `Game` that Bridge can modify and `Card` will use it to adjust the cost accordingly.

I think this will be useful when implementing other expansions as certain cards (e.g. [Peddler](http://wiki.dominionstrategy.com/index.php/Peddler), [Destrier](http://wiki.dominionstrategy.com/index.php/Destrier), etc.) have costs that can vary. These cards can override `get_cost()` to implement whatever logic they need to adjust their costs.

## Mining Village and `multi_play()`

[Mining Village](http://wiki.dominionstrategy.com/index.php/Mining_Village) was one of the most challenging cards to implement. This was due to an interesting case when it is played by Throne Room. Mining Village cannot be trashed twice, and thus players cannot get the +$2 twice. This means, on its second play by Throne Room, Mining Village needs to know if it's already been trashed. However, I could not check to see if the Mining Village card instance was in the trash as there is only one instance of the card, so it is not possible to tell if a Mining Village in the trash was just trashed or if a player trashed a Mining Village on a previous turn. The solution I came up with was to create a new `Card` method called `multi_play()` that Throne Room calls to play a card multiple times. By default, `multi_play()` just calls `play()`, but it can be overridden by derived cards to implement specific behavior. `multi_play()` returns an object that will be passed back into the function on the second call by Throne Room. For Mining Village, `multi_play()` passes a boolean value indicating whether it has been trashed, so it can know whether it should prompt the player to trash itself the second time it is played.

This was the best way I could come up with to solve this problem, but I'm not tied to it if you think of a better way.

evanofslack commented:
Wow this is so freaking cool! I've been periodically checking this intrigue branch and awaiting the completion. 

I'm going to playtest these new cards!

evanofslack commented:
Played a few rounds against big money, super fun!

Gonna drop some comments:

Love the `diplomat.on_attack` method, makes me want to change moat to have the reaction logic inside the card

Nitpick here, we can do an early return here for lurker when no trash and no supply (rare) and shift the indent level to make the rest a bit cleaner.

```python
        if len(supply_action_cards) == 0 and len(trash_action_cards) == 0:
            return
        if len(supply_action_cards) == 0:
        ...
```

For masquerade, before asking player if they want to trash a card from there hand, lets do a check to make sure they have cards in hand (rare but possible)

Another nit for TradingPost with indentation

```python
        if len_hand == 0:
            pass
        elif len_hand == 1:
            player.trash(player.hand.cards[0], game.trash)
        else:
            if len_hand == 2:
                trash_cards = player.hand.cards[:]
            else:
                trash_cards = player.decider.trash_decision(
```

goes to 

```python
        if len_hand == 0:
            return
        elif len_hand == 1:
            trash_cards = [player.hand.cards[0]]
        elif len_hand == 2:
            trash_cards = player.hand.cards[:]
        else:
            trash_cards = player.decider.trash_decision()
            ...

        for card in trash_cards:
            player.trash(card, game.trash)
```

Overall looks great, as always appreciate the work, I know its a bit tedious to do a whole set.

jbw3 commented:
Glad you enjoyed them! I'll make the suggested changes.

jbw3 commented:
I've finished updating `OptimizedBotDecider` and addressed your comments. Let me know if there is anything else I should change. Otherwise, this branch should be good to go!

evanofslack commented on tests/conftest.py:
```diff
@@ -101,7 +106,8 @@ def supply():
         players=[player],
-        expansions=[base_set],
+        expansions=[base_set, intrigue_set],
+        kingdom_cards=[mill, smithy],
```
Are mill and smithy required here (and line 126)? I could have overlooked something.

jbw3 replied:
I'm used them in some of the tests for cards that interact with the supply. I used mill in `test_ironworks,py` and `test_replace.py` and smithy in `test_ironworks.py` and `test_lurker.py`.

evanofslack replied:
Ok thanks for the explanation, good to keep them in then.

evanofslack reviewed (changes requested):
Thanks for the contribution! Killer stuff. Good to merge after that clarifying questions.

+1
